core/src/main/java/org/apache/iceberg/util/PartitionUtil.java

```diff
@@ -39,13 +40,17 @@ private PartitionUtil() {
   }
 
   public static Map<Integer, ?> constantsMap(FileScanTask task, BiFunction<Type, Object, Object> convertConstant) {
-    return constantsMap(task.spec(), task.file().partition(), convertConstant);
-  }
+    PartitionSpec spec = task.spec();
+    StructLike partitionData = task.file().partition();
 
-  private static Map<Integer, ?> constantsMap(PartitionSpec spec, StructLike partitionData,
-                                              BiFunction<Type, Object, Object> convertConstant) {
     // use java.util.HashMap because partition data may contain null values
     Map<Integer, Object> idToConstant = new HashMap<>();
+
+    // add _file
+    idToConstant.put(
+        MetadataColumns.FILE_PATH.fieldId(),
+        convertConstant.apply(Types.StringType.get(), task.file().path()));
```

Comment:
This adds `_file` to the constants map so it is set in records like a partition value.